### PR TITLE
FEXLoader: Allocate the second 4GB of virtual memory when executing 32-bit

### PR DESCRIPTION
**Pull request by Sonicadvance1**
Spurred on by https://github.com/FEX-Emu/FEX/pull/3421. To ensure that applications don't take advantage of
small address wrap around, allocate the second 4GB of virtual memory.

Some context. Linux always reserves the first 16KB of virtual address
space (unless you tinker with some settings which nobody should do).

Example of 32-bit code:
lea eax, [0xffff_0000]
mov ebx, [eax + 0x1_0000]

The address calculated by the mov will wrap around to 0x0 which will
result in SIGSEGV. If FEX messes up zero extensions then it would try to
access 0x1_0000_0000 instead.

This could result in a 32-bit application potentially accessing some FEX
memory instead of crashing.
Add this safety net which will still SIGSEGV and we will be able to see
the crash.

### CODE REVIEW

**Comment by alyssarosenzweig** [comment]
It's really the *second* 4GB we're allocating, no?

**Comment by Sonicadvance1** [comment]
> It's really the _second_ 4GB we're allocating, no?

Maybe it should be phrased that way. As usual my English phrasing is derp.

**Review by alyssarosenzweig — APPROVED** [review]

**Comment by pmatos** [comment]
LGTM - curious if there a reason why you decided to allocate a full 4GB rather than a smaller window?

File: Source/Tools/FEXLoader/FEXLoader.cpp
Comment (neobrain) [inline Source/Tools/FEXLoader/FEXLoader.cpp:427]: Shouldn't this have a comment explaining why we're stealing this particular memory region?

**Comment by Sonicadvance1** [comment]
> LGTM - curious if there a reason why you decided to allocate a full 4GB rather than a smaller window?

Primary to ensure if we mess up adding together two 32-bit values, it still captures the overflow. Bigger safety net.
And allocating 16KB versus 4GB has the same overhead so it doesn't really matter.

**Comment by Amanieu** [comment]
You should reserve the first 64G so that you can use shifted-add addressing modes on AArch64 with 32-bit zero-extended values: `ldr w0, [x1, x2, sxtw #3]` instead of needing a separate 32-bit `add` for zero-extended address calculation.

This will almost always work and return an address in the low 4G, and you can handle the cases where it doesn't by catching the signal and backpatching. See section 3.3 of [my paper](https://pure.manchester.ac.uk/ws/portalfiles/portal/56078084/pldi_16.pdf) for more details.

File: Source/Tools/FEXLoader/FEXLoader.cpp
Comment (Sonicadvance1) [reply Source/Tools/FEXLoader/FEXLoader.cpp:429]: Added a coment.

**Comment by Sonicadvance1** [comment]
> You should reserve the first 64G so that you can use shifted-add addressing modes on AArch64 with 32-bit zero-extended values: `ldr w0, [x1, x2, sxtw #3]` instead of needing a separate 32-bit `add` for zero-extended address calculation.
> 
> This will almost always work and return an address in the low 4G, and you can handle the cases where it doesn't by catching the signal and backpatching. See section 3.3 of [my paper](https://pure.manchester.ac.uk/ws/portalfiles/portal/56078084/pldi_16.pdf) for more details.

Very good information, we can look at increasing the number to 64GB once/if we land that optimization.

**Comment by pmatos** [comment]
> You should reserve the first 64G so that you can use shifted-add addressing modes on AArch64 with 32-bit zero-extended values: `ldr w0, [x1, x2, sxtw #3]` instead of needing a separate 32-bit `add` for zero-extended address calculation.
> 
> This will almost always work and return an address in the low 4G, and you can handle the cases where it doesn't by catching the signal and backpatching. See section 3.3 of [my paper](https://pure.manchester.ac.uk/ws/portalfiles/portal/56078084/pldi_16.pdf) for more details.

I only just noticed this comment and I think it's related to the patch I am working on #3421. Thanks for this.

**Pull request merged**
